### DISCUSSION TITLE
Support for Jupyter Notebooks in code reviews

### DISCUSSION THREAD

**Discussion by gabroo** [community/community#12959] [Pull Requests]
Currently in PRs, when viewing files with the `.ipynb` extension, they are displayed only in the raw JSON representation, which makes it essentially useless for conducting code reviews. The line numbers don't align with the notebook, there's no syntax highlighting, all the metadata is viewable, etc.

It would be amazing to have the files rendered in the familiar notebook format with cells. 

Ideally, it would be nice to also have support for rendering interactive cell outputs (ie, not just static images), but at a minimum, adding comments to notebooks in the cell format should be supported.

Is this on your radar? If so, is there any kind of timeline for this? 

I would assume this a huge pain point for many data science teams that would like to use GitHub for source control, but simply cannot because their go-to file format is unsupported in code reviews.

**Comment by gaow** [comment]
I echo this feature request. Currently I rely on [this 3rd party service](https://www.reviewnb.com/).

**Comment by Tobeyforce** [comment]
Same, as someone doing code reviews for university students (with github classroom) I can't give proper feedback on notebooks.

**Reply by serghiou** [reply]
For an academic setting you may want to consider [Colab](https://colab.research.google.com/). Students can start coding with zero setup, easily import any pre-work you may have done and available on GitHub, and you can comment on any of their code as if commenting in Google Docs.

**Reply by Tobeyforce** [reply]
Yes, but I want to be able to track students part of a classroom, and I want to be able to mix other types of assignments :-) github classroom is fantastic so far, but I'd be very curious to hear how you might have set things up :)

**Comment by therzka** [comment]
@Tobeyforce @gaow @gabroo 

👋 We've just opened up a feature preview for diffing Jupyter Notebooks, please fill out [this form](https://docs.google.com/forms/d/e/1FAIpQLSe7UL0qzaP_rRvnSNzGQgE4YmEJpMENWyDISDBleCz_IQLlFQ/viewform) for access and provide feedback [in this discussion](https://github.com/orgs/community/discussions/37376)

**Reply by gabroo** [reply]
Thank you @therzka!

**Reply by Tobeyforce** [reply]
Thank you! I signed up.

**Reply by ostegm** [reply]
I signed up, but have not gotten access, is this feature still available to preview?

**Reply by ruthappel** [reply]
Hi @therzka, thanks a lot for creating the feature preview – I signed up a few days ago, but don't seem to have gotten access yet. Is it still available?

**Reply by therzka** [reply]
@ruthappel @ostegm this feature is now an official feature preview that you can enable for yourself!  (click on your avatar and then on feature previews)

**Reply by ruthappel** [reply]
Thank you so much @therzka, this is great :)

**Reply by miaojingang** [reply]
@therzka Do you know when the preview will be available in github enterprise?

**Comment by smith-kyle** [comment]
Not sure what the timeline is for this feature to be finished, but in the meantime you can use [GitNotebooks](https://gitnotebooks.com). It's free for private repos.

Just a little friendly competition 😛

**Comment by Paul1911** [comment]
Hey @ruthappel, thank you for taking this up :) I was just wondering - what is the official timeline for going out of testing into production? We could really need it at work, but as we don't have access to feature previews, we can't make use of it atm.

**Comment by nathanneuro** [comment]
any news on this?

**Comment by droumis** [comment]
any update? where is this on the roadmap?

**Comment by princelySid** [comment]
You can enable this in Feature preview, found by clicking your picture in the top right.

**Comment by dchess** [comment]
While the Rich Jupyter Notebook Diffs feature is available in preview, there is still no way to comment on the diffs. Is there a timeline for that feature. Its a real limitation for data teams.

**Comment by rmaddenatcffc** [comment]
GitHub should just acquire GitNotebooks and assimilate it into the product. We need to be able to elegantly preview notebooks and leave comments in reviews. It's 2025 now.

**Comment by droumis** [comment]
bump

**Comment by lauragalera** [comment]
Any updates on when the feature will be out? We also want to suggest changes through comments directly on the code

**Comment by kinano** [comment]
Hello 👋. Why is this feature request not prioritized 3 years in?

**Comment by JacobRSchwartz-AI** [comment]
Bumping for visibility

**Comment by hcjmartin-work** [comment]
+1 bump. 

Genuinely surprised on this one. 

Where are all the Data Engineers? Not on github?

**Comment by Paul1911** [comment]
Just use Marimo notebooks. We're also switching now.

**Comment by amit1rrr** [comment]
Since many of you are asking about commenting on notebook diffs, [ReviewNB](https://www.reviewnb.com/) is built specifically for this. It provides rich notebook diffs and cell-level comments.

**Disclosure**: I built ReviewNB. We've been focused on notebook code reviews on GitHub since 2019. It's free for open source and education, and is used by thousands of repositories. Here are some [popular open source repositories using ReviewNB](https://www.reviewnb.com/#repos).

Of course, it would be great to have this built natively into GitHub, but until that happens I just wanted to point you to a solution that’s available today.

<img width="2470" height="1462" alt="rnb_diff" src="https://github.com/user-attachments/assets/d712e827-4715-4bb3-96c3-5d2a6c6ee84b" />
